cirq/circuits/circuit.py

```diff
@@ -871,6 +874,27 @@ def _can_commute_past(self,
                           operation: ops.Operation) -> bool:
         return not self._moments[moment_index].operates_on(operation.qubits)
 
+    def _validate_op_tree_qids(self, op_tree: ops.OP_TREE) -> None:
+        """Raises an exception if any operation in `op_tree` has qids that don't
+        match its qid shape.
+
+        Args:
+            operation: The operation to validate.
+
+        Raises:
+            ValueError: The operation had qids that don't match its qid shape.
+        """
+        # Cast from Iterable[Operation, Moment] because preserve_moments is
+        # False.
+        for op in cast(Iterable[ops.Operation], ops.flatten_op_tree(op_tree)):
+            if protocols.qid_shape(op) != protocols.qid_shape(op.qubits):
```

Comment:
We can't have operations validate that their qubits are of the correct shape?

Is this forward-looking for passing qutrits into a qubit gate? Is it defense in depth?

Reply:
Gates are verified by GateOperation.  This would catch operations that are unaware of qudits.  E.g. `cirq.ControlledOperation([cirq.LineQid(0, 3)], cirq.GlobalPhaseOperation(1))` doesn't fail but this check catches it.